include/rocksdb/configurable.h

```diff
@@ -0,0 +1,493 @@
+  }
+  virtual ~Configurable() {}
+
+ public:  // Public API methods exposed by Configurable
+  /**
+   *  Attempts to make the local options valid for the specified DB Options
+   * and ColumnFamilyOptions. If necessary, this method can change values of
+   * itself or the DBOptions or the ColumnFamilyOptions (unlike
+   * ValidateOptions). After calling this method, the
+   * DBOptions/ColumnFamilyOptions should be valid. a non-ok Status will be
+   * returned.
+   */
+  Status SanitizeOptions();
+  Status SanitizeOptions(DBOptions& db_opts);
+  Status SanitizeOptions(DBOptions& db_opts, ColumnFamilyOptions& cf_opts);
+
+  /**
+   *  Validates the local options are valid for the specified DB Options
+   * and ColumnFamilyOptions. If the options are not valid,
+   * a non-ok Status will be returned.  This method cannot change its input
+   * arguments or the object itself.
+   */
+  Status ValidateOptions() const;
+  Status ValidateOptions(const DBOptions&) const;
+  Status ValidateOptions(const DBOptions& db_opts,
+                         const ColumnFamilyOptions& cf_opts) const;
+
+  /**
+   *  Prints the configuration object settings to the input Logger
+   */
+  void Dump(Logger* log, const std::string& indent,
+            uint32_t mode = OptionStringMode::kOptionNone) const {
+    DumpOptions(log, indent, mode);
+  }
+  void Dump(Logger* log, uint32_t mode = OptionStringMode::kOptionNone) const {
+    Dump(log, "              ", mode);
+  }
+
+  /**
+   *  Returns the raw pointer of the named options that is used by this
+   * object, or nullptr if this function is not supported.
+   * Since the return value is a raw pointer, the object owns the
+   * pointer and the caller should not delete the pointer.
+   *
+   * Note that changing the underlying options while the object
+   * is currently used by any open DB is undefined behavior.
+   * Developers should use DB::SetOption() instead to dynamically change
+   * options while the DB is open.
+   */
+  template <typename T>
+  const T* GetOptions(const std::string& name) const {
+    return reinterpret_cast<const T*>(GetOptionsPtr(name));
+  }
+  template <typename T>
+  T* GetOptions(const std::string& name) {
+    return reinterpret_cast<T*>(const_cast<void*>(GetOptionsPtr(name)));
+  }
+
+#ifndef ROCKSDB_LITE
+  /**
+   * Configures the options for this extension based on the input parameters.
+   * Returns an OK status if configuration was successful.
+   * Parameters:
+   *   opt_map                 The name/value pair map of the options to set
+   *   opt_str                 String of name/value pairs of the options to set
+   *   input_strings_escaped   True if the strings are escaped (old-style?)
+   *   ignore_unusued_options  If true and there are any unused options,
+   *                           they will be ignored and OK will be returned
+   *   unused_opts             The set of any unused option names from the map
+   */
+  Status ConfigureFromMap(const DBOptions&,
```

Comment:
I believe we have been trying to avoid default arguments for the purpose of less ambiguity, especially if there are multiple overloads. For legacy code, we may have to live with them, but for newly-added code, it is better if we can avoid them.

Reply:
Do you want all default arguments removed, or only from the constructors?  I removed the ones from the constructors, but there are a lot of places that take other defaults (like ignore_unknown_options, input_strings_escaped, delimiter).  I can make new signatures without default arguments, but in this case I am not sure it simplifies anything to add them.

Reply:
For existing interface, we can keep default arguments. For newly-added functions, I think we should avoid them.